controllers/perconaserverformysql_controller.go

```diff
@@ -0,0 +1,333 @@
+	cr *apiv2.PerconaServerForMySQL,
+) error {
+	initImage, err := k8s.InitImage(ctx, r.Client)
+	if err != nil {
+		return errors.Wrap(err, "get init image")
+	}
+
+	err = k8s.EnsureObjectWithHash(ctx, r.Client, cr, mysql.StatefulSet(cr, initImage), r.Scheme)
+	if err != nil {
+		return errors.Wrap(err, "reconcile sts")
+	}
+	err = k8s.EnsureObjectWithHash(ctx, r.Client, cr, mysql.Service(cr), r.Scheme)
+	if err != nil {
+		return errors.Wrap(err, "reconcile svc")
+	}
+	err = k8s.EnsureObjectWithHash(ctx, r.Client, cr, mysql.PrimaryService(cr), r.Scheme)
+	if err != nil {
+		return errors.Wrap(err, "reconcile primary svc")
+	}
+	err = k8s.EnsureObjectWithHash(ctx, r.Client, cr, mysql.UnreadyService(cr), r.Scheme)
+	if err != nil {
+		return errors.Wrap(err, "reconcile unready svc")
+	}
+
+	return nil
+}
+
+func (r *PerconaServerForMySQLReconciler) reconcileOrchestrator(
+	ctx context.Context,
+	cr *apiv2.PerconaServerForMySQL,
+) error {
+	err := k8s.EnsureObjectWithHash(ctx, r.Client, cr, orchestrator.StatefulSet(cr), r.Scheme)
+	if err != nil {
+		return errors.Wrap(err, "reconcile StatefulSet")
+	}
+	err = k8s.EnsureObjectWithHash(ctx, r.Client, cr, orchestrator.Service(cr), r.Scheme)
+	if err != nil {
+		return errors.Wrap(err, "reconcile Service")
+	}
+
+	return nil
+}
+
+func (r *PerconaServerForMySQLReconciler) reconcileReplication(
+	ctx context.Context,
+	cr *apiv2.PerconaServerForMySQL,
+) error {
+	if err := reconcileReplicationPrimaryPod(ctx, r.Client, cr); err != nil {
+		return errors.Wrap(err, "reconcile primary pod")
+	}
+	if err := reconcileReplicationSemiSync(ctx, r.Client, cr); err != nil {
+		return errors.Wrap(err, "reconcile semi-sync")
+	}
+
+	return nil
+}
+
+func reconcileReplicationPrimaryPod(
+	ctx context.Context,
+	cl k8s.APIListPatcher,
+	cr *apiv2.PerconaServerForMySQL,
+) error {
+	pods, err := k8s.PodsByLabels(ctx, cl, mysql.MatchLabels(cr))
+	if err != nil {
+		return errors.Wrap(err, "get MySQL pod list")
+	}
+
+	host := orchestrator.APIHost(orchestrator.ServiceName(cr))
+	primary, err := orchestrator.ClusterPrimary(ctx, host, cr.ClusterHint())
+	if err != nil {
+		return errors.Wrap(err, "get cluster from orchestrator")
+	}
+	primaryAlias := primary.Alias()
+
+	for i := range pods {
+		pod := &pods[i]
+		if pod.GetLabels()[apiv2.MySQLPrimaryLabel] == "true" {
+			if pod.Name == primaryAlias {
+				// primary is not changed
+				return nil
+			}
+
+			patch := client.StrategicMergeFrom(pod)
+			k8s.RemoveLabel(pod, apiv2.MySQLPrimaryLabel)
+			if err := cl.Patch(ctx, pod, patch); err != nil {
+				return errors.Wrap(err, "remove label from old primary pod")
+			}
+
+			break
+		}
+	}
+
+	for i := range pods {
+		pod := &pods[i]
+		if pods[i].Name == primaryAlias {
+			patch := client.StrategicMergeFrom(pod)
+			k8s.AddLabel(pod, apiv2.MySQLPrimaryLabel, "true")
+			if err := cl.Patch(ctx, pod, patch); err != nil {
+				return errors.Wrap(err, "add label to new primary pod")
+			}
+
+			break
+		}
+	}
```

Comment:
why don't we just `Get` the pod with name `primaryAlias`?

Reply:
we still need the list of pods here to remove label from old primary